/*========================== begin_copyright_notice ============================

Copyright (C) 2026 Intel Corporation

SPDX-License-Identifier: MIT

============================= end_copyright_notice ===========================*/

/*========================== begin_copyright_notice ============================

Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
See https://llvm.org/LICENSE.txt for license information.
SPDX-License-Identifier: Apache-2.0 with LLVM-exception

============================= end_copyright_notice ===========================*/

This comes from patches of LLVM 14 (copy-pasted from there)
diff --git a/llvm/lib/Transforms/Scalar/LoopUnrollPass.cpp b/llvm/lib/Transforms/Scalar/LoopUnrollPass.cpp
index 7034e595a4..e7b507afbf 100644
--- a/llvm/lib/Transforms/Scalar/LoopUnrollPass.cpp
+++ b/llvm/lib/Transforms/Scalar/LoopUnrollPass.cpp
@@ -1131,7 +1131,7 @@ void llvm::computeUnrollCount(Loop *L, const TargetTransformInfo &TTI,
   // found it unprofitable, we'll never chose to bounded unroll.
   LLVM_DEBUG(dbgs().indent(1) << "Trying upper-bound unroll...\n");
   if (!TripCount && MaxTripCount && (UP.UpperBound || MaxOrZero) &&
-      MaxTripCount <= UP.MaxUpperBound) {
+      MaxTripCount < std::max(16U, UP.MaxUpperBound)) {
     if (auto UnrollFactor = shouldFullUnroll(L, TTI, DT, SE, EphValues,
                                              MaxTripCount, UCE, UP)) {
       UP.Count = *UnrollFactor;
@@ -1175,7 +1175,7 @@ void llvm::computeUnrollCount(Loop *L, const TargetTransformInfo &TTI,
   }
 
   // Don't unroll a small upper bound loop unless user or TTI asked to do so.
-  if (MaxTripCount && !UP.Force && MaxTripCount <= UP.MaxUpperBound) {
+  if (MaxTripCount && !UP.Force && MaxTripCount < std::max(16U, UP.MaxUpperBound)) {
     LLVM_DEBUG(dbgs().indent(2) << "Not runtime unrolling: max trip count "
                                 << MaxTripCount << " is small (<= "
                                 << UP.MaxUpperBound << ") and not forced.\n");

--
2.43.0
